File: java/src/Simulacion.java
package simulacion;

import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

/**
 * Modela la simulación de un día en el cajero Banpato.
 *
 * @author  Luis Clemente Zúñiga Flores
 */
public class Simulacion {

//    Campos
    /**
     * Cantidad de eventos a imprimir
     */
    public static int eventosAImprimir = 0;

    /**
     * Representa la cantidad de eventos
     */
    public static int cantidadDeEventos = 0;

    /**
     * Representa la cantidad de llegadas al cajero
     */
    public static int cantidadDeLlegadas = 0;
    /**
     * Guarda si hay o no abastecimiento en tránsito
     */
    public static boolean abastecimientoEnTransito = false;

    /**
     * Guarda la cantidad de reabastecimientos
     */
    public static int cantidadDeAbastecimientos;

    /**
     * Guarda el total de dinero reabastecido
     */
    public static double cantidadReabastecida;

    /**
     * Cantidad de disposición promedio cajero.cantidadAbastecida/clientes
     * atendidos
     */
    public static double disposicionPromedio;

    /**
     * Guarda el tiempo de espera calcular (tiempo de espera/clientes) atendidos
     */
    public static double tiempoDeEspera;

    /**
     * Guarda el tiempo de servicio calcular (tiempo de servicio/clientes)
     * atendidos
     */
    public static double tiempoDeServicio;

    /**
     * Guarda el tiempo fuera de servicio
     */
    public static double tiempoFueraDeServicio;

    /**
     * Temporal de fuera de servicio
     */
    public static double temporalDeFueraDeServicio;

    /**
     * Guarda la cantidad de usuarios insatisfechos.
     */
    public static int usuariosInsatisfechos;

    /**
     * Guarda la cantidad de usuarios atendidos
     */
    public static int usuariosAtendidos;
    /**
     * Guarda la hora actual de la simulación.
     */
    public static double reloj = 0;

    /**
     * Cuenta la cantidad de usuarios formados
     */
    public static int usuariosEnSistema;

    /**
     * Guarda el total de las disposiciones
     */
    public static double montoDispuestoTotal = 0;
    /**
     * Instancia de un cajero automático.
     */
    public static Cajero cajero = new Cajero(500000, 0);

    /**
     * Tabla de eventos futuros
     */
    public static LinkedList<Evento> tablaDeEventosFuturos = new LinkedList<>();

    /**
     * Guarda los usuarios que llegan en una cola
     */
    public static Queue<Usuario> colaUsuarios = new LinkedList<>();

    public static Stack<Usuario> pilaUsuarios = new Stack<>();

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        Writer.writer("SIMULACIÓN DE UN CAJERO AUTOMÁTICO.\nAlumno: Luis Clemente Zúñiga Flores.\nProfesor: Martín Nicolás Silva Arredondo.\nMateria: Simulación.\n-------------------------------------------------------------------------------\n");
        tablaDeEventosFuturos.add(new Evento(1, 32400));
        tablaDeEventosFuturos.add(new Evento(2, 64800));
        do {
            Collections.sort(tablaDeEventosFuturos, Comparator.comparingDouble(Evento::getTiempoDeOcurrencia));
            monitorDeEventos();
            tablaDeEventosFuturos.remove();
        } while (reloj < 64800 && tablaDeEventosFuturos.size() > 0);

        resultadosFinales();

    }

    /**
     * Monitorea los eventos y de acuerdo a su tipo, ejecuta el método adecuado.
     */
    public static void monitorDeEventos() {
        int tipo = tablaDeEventosFuturos.getFirst().getTipoDeEvento();
        double tiempo = tablaDeEventosFuturos.getFirst().getTiempoDeOcurrencia();
        switch (tipo) {

            case 1:
                abrir(tiempo);
                break;
            case 2:
                cerrar(tiempo);
                break;
            case 3:
                llegada(tiempo);
                break;
            case 4:
                atencion(tiempo);
                break;
            case 5:
                salida(tiempo);
                break;
            case 6:
                abastecimiento(tiempo);
                break;

        }
    }

    /**
     * Establece los parámetros iniciales del cajero.
     *
     * @param hora indica la hora en que se abre el cajero
     */
    public static void abrir(double hora) {
//        Actualizar el reloj
        reloj = hora;
//        Aumenta contador de eventos
        cantidadDeEventos++;
//        poner estatus abierto
        cajero.setAbierto(true);
        cajero.setEnLinea(true);
        cajero.setOcupado(false);
//        Generar evento llegada el el tiempo reloj + Aleatorio de tiempo entre llegadas
        double eventoLlegada = reloj + Aleatorio.exponencial(500);
        tablaDeEventosFuturos.add(new Evento(3, eventoLlegada));
//        colaUsuarios.add(new Usuario(eventoLlegada));
//        Desplegar evento en su caso
        desplegarEvento(1, reloj);

        System.out.println("cajero abierto: " + cajero.isAbierto() + "\nCajero en línea: " + cajero.isEnLinea() + "\nCajero ocupado: " + cajero.isOcupado());

        if (!cajero.isOcupado()) {
            System.out.println("El cajero no está ocupado.");
        }

    }

    /**
     * Establece el cierre del cajero.
     *
     * @param hora indica la hora de cierre del cajero
     */
    public static void cerrar(double hora) {
//        Actualizar reloj
        reloj = hora;
//        Aumentar contador de número de eventos
        cantidadDeEventos++;
//        Poner estatus cerrado
        cajero.setAbierto(false);
//        Desplegar evento en su caso
        desplegarEvento(2, reloj);
    }

    /**
     * Establece la llegada de un cliente y prepara lo necesario para su
     * atención.
     *
     * @param hora indica la llegada del un cliente
     */
    public static void llegada(double hora) {
//        Actualizar reloj
        reloj = hora;
//        Incrementar contador de número de eventos
        cantidadDeEventos++;
//        Incrementar número de llegadas
        cantidadDeLlegadas++;
        if (cajero.isAbierto()) {
            double eventoLlegada = reloj + Aleatorio.exponencial(500);
//            Si estatus abierto generar evento llegada en el tiempo Reloj + Aleatorio de tiempo entre llegadas
            tablaDeEventosFuturos.add(new Evento(3, eventoLlegada));

        }
//        Si cajero en línea y estatus abierto
        if (cajero.isEnLinea() && cajero.isAbierto()) {
//            Si hay menos de 8 usuarios en el sistema
            if (usuariosEnSistema < 8) {
//                Agregar cliente a la lista alimentando número de llegada y tiempo de llegada                                          
                colaUsuarios.add(new Usuario(reloj));
//                incrementar número de clientes en el sistema
                usuariosEnSistema++;
//                Si cajero desocupado generar evento inmediato de inicio de atención
                if (!cajero.isOcupado()) {
                    tablaDeEventosFuturos.add(new Evento(4, reloj));
                }

            } else {
//                Si no aumentar insatisfechos
                usuariosInsatisfechos++;
            }

        } else {
//            Sino aumentar insatisfechos
            usuariosInsatisfechos++;
        }
//        Desplegar evento en su caso
        desplegarEvento(3, reloj);

    }

    /**
     * Ejecuta el algoritmo de atención al usuario
     *
     * @param hora indica la hora de llegada del usuario
     */
    public static void atencion(double hora) {
//        Actualizar reloj
        reloj = hora;
//        incrementar contador de número de eventos
        cantidadDeEventos++;
//        poner cajero ocupado
        cajero.setOcupado(true);
//Alimentar tiempo de inicio de atención del cliente e incrementar insatisfechos si esperó más de 5 mins
        colaUsuarios.peek().setHoraInicioDeAtencion(reloj);
        if ((colaUsuarios.peek().getHoraInicioDeAtencion() - colaUsuarios.peek().getHoraDeLlegada()) / 60 > 5) {
            usuariosInsatisfechos++;
        }
//        Acumular tiempo de espera
        colaUsuarios.peek().setTiempoDeEspera(colaUsuarios.peek().getHoraInicioDeAtencion() - colaUsuarios.peek().getHoraDeLlegada());
//        Generar evento salida en el tiempo reloj + Aleatorio de tiempo de servicio
        double tiempoDeServicio = reloj + Aleatorio.exponencial(120);
        tablaDeEventosFuturos.add(new Evento(5, tiempoDeServicio));

//        Alimentar el monto solicitado con aleatorio de monto de disposición * 100
        colaUsuarios.peek().setCantidadSolicitada(Aleatorio.uniformeParams(0, 50) * 100);

        if (colaUsuarios.peek().getCantidadSolicitada() < cajero.getCantidadActual()) {
            colaUsuarios.peek().setCantidadRecibida(colaUsuarios.peek().getCantidadSolicitada());
            //Alimentar monto dispuesto con el monto solicitado o lo que quede incrementando insatisfechos
        } else if (colaUsuarios.peek().getCantidadSolicitada() > cajero.getCantidadActual()) {
            colaUsuarios.peek().setCantidadRecibida(cajero.getCantidadActual());
            usuariosInsatisfechos++;
        }
//        Desplegar evento en su caso
        desplegarEvento(4, reloj);
    }

    /**
     * Ejecuta el algoritmo de salida del cliente
     *
     * @param hora indica la hora de salida del cliente.
     */
    public static void salida(double hora) {
//        Actualizar reloj
        reloj = hora;
//Incrementar contador de número de eventos
        cantidadDeEventos++;
//alimentar tiempo de salida
        colaUsuarios.peek().setHoraDeSalida(reloj);
//acumular tiempo de servicio
        colaUsuarios.peek().setTiempoDeServicio(colaUsuarios.peek().getHoraDeSalida() - colaUsuarios.peek().getHoraInicioDeAtencion());
//        Decrementar monto dispuesto del efectivo
        cajero.disposicionDeEfectivo(colaUsuarios.peek().getCantidadRecibida());
//        Acumular monto dispuesto
        montoDispuestoTotal += colaUsuarios.peek().getCantidadRecibida();
//        Eliminar cliente de la lista
        pilaUsuarios.add(colaUsuarios.poll());
//Decrementar número de clientes en el sistema
        usuariosEnSistema--;
//Incrementar número de usuarios atendidos
        usuariosAtendidos++;
//poner cajero desocupado
        cajero.setOcupado(false);
//si efectivo < punto de reorden y no hay abastecimiento en tránsito
        if (cajero.getCantidadActual() <= cajero.getPuntoDeReorden() && !abastecimientoEnTransito) {
//Generar evento abastecimiento en el tiempo reloj + Aleatorio de tiempo de abastecimiento
            double tiempoAbastecimiento = reloj + Aleatorio.exponencial(600);
            tablaDeEventosFuturos.add(new Evento(6, tiempoAbastecimiento));
            abastecimientoEnTransito = true;
        }
        if (cajero.getCantidadActual() == 0) {
//            Poner cajero fuera de línea y registrar tiempo en que se pone off-line
            cajero.setEnLinea(false);
            temporalDeFueraDeServicio = reloj;
//Aumentar insatisfechos con el número de clientes en el sistema
            usuariosInsatisfechos += colaUsuarios.size();
//Vaciar la lista de clientes 
            colaUsuarios.removeAll(colaUsuarios);
            usuariosEnSistema = 0;

        }

//        si número de clientes en sistema > 0
        if (usuariosEnSistema > 0) {
//           Generar evento inmediato de inicio de atención
            tablaDeEventosFuturos.add(new Evento(4, reloj));
        }

//        desplegar evento en su caso
        desplegarEvento(5, reloj);

    }

    /**
     * Ejecuta el algoritmo de abastecimiento de efectivo, en caso de que el
     * dinero disponible sea menor al punto de reorden.
     *
     * @param hora indica la hora de abastecimiento.
     */
    public static void abastecimiento(double hora) {
//        Actualizar reloj
        reloj = hora;
//incrementar contador de número de eventos
        cantidadDeEventos++;
//agregar a efectivo lo que le falta para la cantidad fija de abastecimiento (fija - efectivo)
        double tempActual = cajero.getCantidadActual();
        cajero.setCantidadActual(cajero.getCantidadInicial());

        cajero.setCantidadAbastecida(cajero.getCantidadInicial() - tempActual);
//Acumular lo abastecido
        cantidadReabastecida += cajero.getCantidadAbastecida();
        System.out.println("Cantidad reabastecida: " + cantidadReabastecida);
        cantidadDeAbastecimientos++;

        if (!cajero.isEnLinea()) {
            cajero.setEnLinea(true);
            cajero.setTiempoOffLine(reloj - temporalDeFueraDeServicio);
        }
//        Poner que no hay abastecimiento en tránsito
        abastecimientoEnTransito = false;
//Desplegar evento en su caso
        desplegarEvento(6, reloj);

    }

    /**
     * Despliega el evento dependiendo de su tipo.
     *
     * @param evento indica el tipo de evento
     * @param hora indica la hora de ocurrencia del evento.
     */
    public static void desplegarEvento(int evento, double hora) {

        String Evento = "";
        switch (evento) {
            case 1:
                Evento = "Abrir banco";
                break;
            case 2:
                Evento = "Cerrar banco";
                break;
            case 3:
                Evento = "Llegada de cliente";
                break;
            case 4:
                Evento = "Atención de cliente";
                break;
            case 5:
                Evento = "Salida de cliente";
                break;
            case 6:
                Evento = "Abastecimiento de efectivo";
                break;
        }

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss");
        LocalTime time = LocalTime.MIN.plusSeconds((long) reloj);

        if (eventosAImprimir < 20) {

            String even = String.format("\nHora: %s\nEvento: %s\nClientes atendidos: %s\nClientes insatisfechos: %s\nClientes en este momento: %s\nDinero en el cajero: %s\n-------------------------------------------------------------------------------\n", formatter.format(time), Evento, usuariosAtendidos, usuariosInsatisfechos, usuariosEnSistema, Writer.varo(cajero.getCantidadActual()));
            Writer.writer("Evento a desplegar: " + ++eventosAImprimir + "\n");
            Writer.writer(even);

        }

        System.out.printf("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\nHora: %s\nEvento: %s\nClientes atendidos: %s\nClientes insatisfechos: %s\nClientes en este momento: %s\nDinero en el cajero: %s\n+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n", formatter.format(time), Evento, usuariosAtendidos, usuariosInsatisfechos, usuariosEnSistema, Writer.varo(cajero.getCantidadActual()));

    }

    /**
     * Imprime los resultados finales de la simulación.
     */
    public static void resultadosFinales() {
        double disposicion = 0;
        double tiempoPromedio = 0;
        double tiempoServicioPromedio = 0;

        for (Usuario u : pilaUsuarios) {
            double espera = u.getHoraInicioDeAtencion() - u.getHoraDeLlegada();
            disposicion += u.getCantidadRecibida();
            tiempoPromedio += espera;
            tiempoServicioPromedio += u.getTiempoDeServicio();
        }

        disposicion /= pilaUsuarios.size();
        tiempoPromedio /= pilaUsuarios.size();
        tiempoServicioPromedio /= pilaUsuarios.size();

        String result = String.format("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\nRESULTADOS FINALES:\n Llegadas de clientes: %s\nClientes atendidos: %s\nClientes insatisfechos: %s\nNumero de reabastecimientos: %s\nCantidad reabastecida: %s\nCantidad de disposición promedio: %s\nTiempo de Servicio Promedio: %s segundos\nTiempo de espera promedio: %s\nTiempo fuera de servicio del cajero: %s segundos\n+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n", cantidadDeLlegadas, usuariosAtendidos, usuariosInsatisfechos, cantidadDeAbastecimientos, Writer.varo(cantidadReabastecida), Writer.varo(disposicion), tiempoServicioPromedio, tiempoPromedio, cajero.getTiempoOffLine());
        Writer.writer(result);

        System.out.printf("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\nLlegadas de clientes: %s\nClientes atendidos: %s\nClientes insatisfechos: %s\nNumero de reabastecimientos: %s\nCantidad reabastecida: %s\nCantidad de disposición promedio: %s\nTiempo de Servicio Promedio: %s\nTiempo de espera promedio: %s segundos\nTiempo fuera de servicio del cajero: %s segundos\n+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n", cantidadDeLlegadas, usuariosAtendidos, usuariosInsatisfechos, cantidadDeAbastecimientos, Writer.varo(cantidadReabastecida), Writer.varo(disposicion), tiempoServicioPromedio, tiempoPromedio, cajero.getTiempoOffLine());

    }

}

File: python/simulacion.py
import time
import math, random

class Aleatorio:
    
    def uniforme(a: float, b: float):
        x = random.uniform(a, b)
        return x
    
    def exponencial(m: float):
        y = random.uniform(0,1)
        x = -1 * m * math.log(1-y)
        return x
    
    def erlang(k: int, m: float):
        x = 0
        me = m / k
        for i in range(k):
            x += exponencial(me)
        return x
    
    def normal(m: float, v: float):
        suma = 0
        for i in range(12):
            suma += random.uniform(0,1)
        z = suma - 6
        x = (math.sqrt(v) * z) + m
        return x
    
    def geometrica(p: float):
        x = 1
        f = p
        s = f
        y = random.uniform(0,1)
        print (y)
        while(y > s):
            f = f * (1 - p)
            s += f
            x += 1
        return x
    
    def poisson (lam: float):
        x = 0
        f = math.exp(-1*lam)
        s = f
        y = random.uniform(0,1)
        while(y > s):
            f = f * lam/(x+1)
            s += f
            x += 1
        return x
    
    def binomial(n: int, p: float):
        x = 0
        f = math.exp(n * math.log(1-p))
        s = f
        y = random.uniform(0,1)
        while(y > s):
            f = f*p*(n-x)/((x+1)*(1-p))
            s += f
            x += 1
        return x
    
    def Aleatoria(tipo: int, par1: float, par2: float):
        x = 0
    
        if tipo == 0:
            x = poisson(par1)
        elif tipo == 1:
            x = erlang(tipo, par1)
        elif tipo == 2:
            x = erlang(tipo, par1)
        elif tipo == 3:
            x = erlang(tipo, par1)
        elif tipo == 4:
            x = erlang(tipo, par1)
        elif tipo == 5:
            x = erlang(tipo, par1)
        elif tipo == 6:
            x = erlang(tipo, par1)
        elif tipo == 7:
            x = erlang(tipo, par1)
        elif tipo == 8:
            x = erlang(tipo, par1)
        elif tipo == 9:
            x = erlang(tipo, par1)
        elif tipo == 10:
            x = normal(par1, par2)
        elif tipo == 11:
            x = binomial(par1, par2)
        elif tipo == 12:
            x = uniforme(par1, par2)
        elif tipo == 13:
            x = geometrica(par1)
        elif tipo > 13 or tipo < 0:
            print("Opcion no valida")
        return x
    
    if __name__ == '__main__':
        print()


class simulacion:

    lista_eventos = []
    llegadas = 0
    cuenta_eventos = 0
    cuenta_rellenar = 0
    rellenar_cajero = 0
    clientes_atendidos = 0
    clientes_insatisfechos=0
    clientes_sistema = 0
    reorden = 20000
    cantidad_fija = 50000
    dinero_cajero = 50000
    reloj = 0
    tiempo_offline = 0
    tiempo_offline_total = 0
    dinero_solicitado = Aleatorio.uniforme(0,50)*100
    numero_de_abastecimientos = 0
    cantidad_abastecida_total = 0
    acumulado_servicio = 0
    mondisp = 0
    cajero_ocupado = False
    cajero_en_linea = True
    cajero_abierto = False
    secuencial = []
    usuarios = []

    def abrir_banco(self,r):
        self.reloj = r
        self.cuenta_eventos+=1
        self.cajero_abierto= True
        self.lista_eventos.append([self.reloj + Aleatorio.exponencial(120),"llegar"])
        self.evenToString(self,"abrir",self.reloj)

    def cerrar_banco(self,r):
        self.reloj = r
        self.cuenta_eventos+=1
        self.cajero_abierto = False
        self.evenToString(self,"cerrar",self.reloj)

    def cliente_llegada(self,r):
        self.reloj = r
        self.cuenta_eventos+=1
        self.llegadas+=1
        if (self.cajero_abierto):
            self.lista_eventos.append([self.reloj + Aleatorio.exponencial(120),"llegar"])
        if(self.cajero_en_linea and self.cajero_abierto):
            if(self.clientes_sistema < 8):
                self.secuencial.append([self.reloj,0,0,0,0])
                self.clientes_sistema+=1
                if(not self.cajero_ocupado):
                    self.lista_eventos.append([self.reloj,"atender"])
            else:
                self.clientes_insatisfechos+=1
        else:
            self.clientes_insatisfechos+=1
        self.evenToString(self,"LLegada de cliente",self.reloj)

    def atencion_de_cliente(self,r):
        self.reloj = r
        self.cuenta_eventos+=1
        self.cajero_ocupado = True
        self.secuencial[0][1] = self.reloj 
        if(self.secuencial[0][1]-self.secuencial[0][0] > (5*60)):
            self.clientes_insatisfechos+=1
        self.lista_eventos.append([self.reloj + Aleatorio.exponencial(60),"salida"])
        self.secuencial[0][3] = Aleatorio.uniforme(0,50)*100
        if(self.secuencial[0][3] < self.dinero_cajero):
            self.secuencial[0][4] = self.secuencial[0][3]
        elif(self.secuencial[0][3] > self.dinero_cajero):
            self.secuencial[0][4] = self.dinero_cajero
            self.clientes_insatisfechos+=1
        self.evenToString(self,"Atención de cliente",self.reloj)

    def salida_cliente(self,r):
        self.reloj = r
        self.cuenta_eventos+=1
        self.secuencial[0][2] = self.reloj
        self.acumulado_servicio += self.secuencial[0][2]-self.secuencial[0][1]
        self.dinero_cajero -= self.secuencial[0][4]
        self.mondisp+= self.secuencial[0][4]
        self.usuarios.append(self.secuencial.pop(0))
        self.clientes_sistema-=1
        self.clientes_atendidos+=1
        self.cajero_ocupado = False
        if(self.dinero_cajero <= self.reorden):
            self.lista_eventos.append([self.reloj + Aleatorio.exponencial(120), "abastecer"])
        if(self.dinero_cajero == 0):
            self.cajero_en_linea = False
            self.tiempo_offline = self.reloj
            self.clientes_insatisfechos+= len(self.secuencial)          
            self.secuencial.clear()
            self.clientes_sistema = 0

        if (self.clientes_sistema > 0):
            self.lista_eventos.append([self.reloj,"atender"])

        self.evenToString(self,"salida",self.reloj)

    def abastece_cajero(self,r):
        self.reloj = r
        self.cuenta_eventos += 1
        self.cantidad_abastecida_total+= self.cantidad_fija - self.dinero_cajero
        self.dinero_cajero = self.cantidad_fija
        self.numero_de_abastecimientos+=1
        if(not self.cajero_en_linea):
            self.cajero_en_linea = True
            self.tiempo_offline = self.reloj - self.tiempo_offline
            self.tiempo_offline_total+= self.tiempo_offline
        self.evenToString(self,"abastecimiento",self.reloj)

    def eventoListener(self):
        evento = self.lista_eventos.pop()
        if(evento[1] == "abrir_banco"):
            self.abrir_banco(self,evento[0])
        elif(evento[1] == "cerrar_banco"):
            self.cerrar_banco(self,evento[0])
        elif(evento[1] == "llegar"):
            self.cliente_llegada(self,evento[0])
        elif(evento[1] == "atender"):
            self.atencion_de_cliente(self,evento[0])
        elif(evento[1] == "salida"):
            self.salida_cliente(self,evento[0])
        elif(evento[1] == "abastecer"):
            self.abastece_cajero(self,evento[0])

    def evenToString(self,evento, tiempo):
        dsp = '''
        Hora: {}
        Evento: {}
        LLegada de clientes: {}
        Clientes Atendidos: {}
        Clientes Insatisfechos: {}
        Clientes en este Momento: {}
        Dinero en el cajero: {}
        '''.format(time.strftime("%H:%M:%S", time.gmtime(tiempo)),evento,self.llegadas,self.clientes_atendidos,self.clientes_insatisfechos,self.clientes_sistema,self.dinero_cajero)
        print(dsp)


    def totales(self):
        dispromedio = 0
        tiempespera = 0
        for i in self.usuarios:
            dispromedio+=i[3]
            tiempespera+=(i[1]-i[0])
        dispromedio/=len(self.usuarios)
        tiempespera/=len(self.usuarios)
        total = '''
        Llegadas de clientes: {}
        Clientes atendidos: {}
        Clientes Insatisfechos: {}
        Número de reabastecimientos: {}
        Cantidad reabastecida: {}
        Cantidad de disposición promedio: {}
        Tiempo de Servicio Promedio: {}
        Tiempo de Espera promedio: {}
        Tiempo fuera de servicio del cajero: {}            
        '''.format(self.llegadas,self.clientes_atendidos,self.clientes_insatisfechos,self.numero_de_abastecimientos,self.cantidad_abastecida_total,dispromedio,self.acumulado_servicio/len(self.usuarios),tiempespera,self.tiempo_offline_total)
        print(total)
        


s = simulacion
s.lista_eventos.append([32400,"abrir_banco"])
s.lista_eventos.append([64800,"cerrar_banco"])
while(s.reloj <= 64800 and len(s.lista_eventos) > 0):
    s.lista_eventos = sorted(s.lista_eventos, reverse = True)
    s.eventoListener(s)
s.totales(s)















